tests/v2/validation/upgrade/kubernetes.go

```diff
@@ -26,23 +33,130 @@ const (
 )
 
+// testUpgradeLocalCluster is a function to test upgrading a local cluster.
+func testUpgradeLocalCluster(t *testing.T, client *rancher.Client, clusterName, versionToUpgrade string, isLatestVersion bool) {
+	clusterMeta, err := clusters.NewClusterMeta(client, clusterName)
+	require.NoError(t, err)
+	require.NotNilf(t, clusterMeta, "Couldn't get the cluster meta")
+	logrus.Infof("[%v]: Provider is: %v, Hosted: %v, Imported: %v , Local: %v", clusterName, clusterMeta.Provider, clusterMeta.IsHosted, clusterMeta.IsImported, clusterMeta.IsLocal)
+
+	initCluster, err := bundledclusters.NewWithClusterMeta(clusterMeta)
+	require.NoError(t, err)
+
+	cluster, err := initCluster.Get(client)
+	require.NoError(t, err)
+
+	versions, err := cluster.ListAvailableVersions(client)
+	require.NoError(t, err)
+	logrus.Infof("[%v]: Available versions for the cluster: %v", clusterName, versions)
+
+	version := getVersion(t, clusterName, versions, isLatestVersion, versionToUpgrade)
+	require.NotNilf(t, version, "Couldn't get the version")
+	logrus.Infof("[%v]: Selected version: %v", clusterName, *version)
+
+	updatedCluster, err := cluster.UpdateKubernetesVersion(client, version)
+	require.NoError(t, err)
+
+	logrus.Infof("[%v]: Validating sent update request for kubernetes version of the cluster", clusterName)
+	validateKubernetesVersions(t, client, updatedCluster, version, isCheckingCurrentCluster)
+
+	logrus.Infof("[%v]: Waiting cluster to be upgraded and ready", clusterName)
+	err = waitUntilLocalStable(client, clusterName)
+	require.NoError(t, err)
+}
+
+// testUpgradeSingleCluster is a function to test upgrading a single cluster.
+func testUpgradeSingleCluster(t *testing.T, client *rancher.Client, clusterName, versionToUpgrade, psact string, isLatestVersion bool) {
+	clusterMeta, err := clusters.NewClusterMeta(client, clusterName)
+	require.NoError(t, err)
+	require.NotNilf(t, clusterMeta, "Couldn't get the cluster meta")
+	logrus.Infof("[%v]: Provider is: %v, Hosted: %v, Imported: %v , Local: %v", clusterName, clusterMeta.Provider, clusterMeta.IsHosted, clusterMeta.IsImported, clusterMeta.IsLocal)
+
+	initCluster, err := bundledclusters.NewWithClusterMeta(clusterMeta)
+	require.NoError(t, err)
+
+	cluster, err := initCluster.Get(client)
+	require.NoError(t, err)
+
+	versions, err := cluster.ListAvailableVersions(client)
+	require.NoError(t, err)
+	logrus.Infof("[%v]: Available versions for the cluster: %v", clusterName, versions)
+
+	version := getVersion(t, clusterName, versions, isLatestVersion, versionToUpgrade)
+	require.NotNilf(t, version, "Couldn't get the version")
+	logrus.Infof("[%v]: Selected version: %v", clusterName, *version)
+
+	updatedCluster, err := cluster.UpdateKubernetesVersion(client, version)
+	require.NoError(t, err)
+
+	logrus.Infof("[%v]: Validating sent update request for kubernetes version of the cluster", clusterName)
+	validateKubernetesVersions(t, client, updatedCluster, version, isCheckingCurrentCluster)
```

Comment:
the `isCheckingCurrentCluster` is confusing to me, line 100 has it set to true, but like 107 sets it to false. Why?

Reply:
We can remove that. I am not sure why it's like that, but it seems like the idea was that after `WaitClusterToBeUpgraded` is ran, that gets ran again to ensure it's done. But we can just remove it IMO and it'll be fine.

Reply:
Resolved.